### PR TITLE
Correct possibly stale cache from network

### LINKED ISSUE #8581: cache-min & unknown version
Currently, if npm is trying to install a package not yet in the npm cache, it will reach out to the registry and grab the package, even with `cache-min=Infinity`. This is desirable.

This allows high values for cache-min to work well, but a problem occurs in the following scenario: 

1. package A in cache, knows all versions up to v1.0.0.
2. package A releases version 1.1.0
3. package B not in cache, gets downloaded
4. package B requests package A ^1.1.0.
5. `npm install --cache-min=Infinity`
6. npm errors with a "no matching version found" error, listing all versions up to 1.0.0.

To resolve this, one has to update the cache's version info e.g. `npm info packagename` or lower the `cache-min` setting. This makes use of high `cache-min` values particularly error prone, especially on first installs.

It would be good if unknown versions were treated the same as unknown package names, i.e. npm should fall back to checking the remote registry before reporting "no compatible version found".

### PR DESCRIPTION
**Pull request by Zirak**
When a package is fetched from the cache which cannot satisfy the version requirements, an attempt to fetch it fron the network is made.

Fix #8581.

I'm in the middle of figuring out how to properly test this: it requires setting up the cache and  publishing a new version (one way or another). I tested it logically using _science_*.

\* Science in this case being manually editing the `.cache.json` file.

### CODE REVIEW

**Comment by timoxley** [comment]
@Zirak nice, thank you for looking into this.

File: lib/fetch-package-metadata.js [hunk 1]
```diff
@@ -149,6 +149,16 @@ function fetchNamedPackageData (dep, next) {
         }
       }
 
+      // We didn't manage to find a compatible version
+      // If this package was requested from cache, force hitting the network
+      if (pkg._cached) {
+        log.silly('fetchNamedPackageData', 'No valid target from cache, forching network')
```
Comment (ralt) [inline lib/fetch-package-metadata.js]: Typo ("forching")

Comment (Zirak) [inline lib/fetch-package-metadata.js]: @ralt Thanks, fixed.

**Comment by iarna** [comment]
Thanks for putting this all together, we landed this today in 3.9.0!

**Comment by iarna** [comment]
FYI, I rewrote your test a bit to not use a fixture folder (we're moving away from that). You can see the rewritten test here: https://github.com/npm/npm/blob/111ae3ec366ece7ebcf5988f5bc2a7cd70737dfe/test/tap/retry-on-stale-cache.js

**Pull request closed without merging**